scylla-cql/src/frame/response/result.rs

```diff
@@ -603,99 +603,104 @@ pub enum Result {
-                0x0014 => TinyInt,
-                0x0015 => Duration,
-                0x0020 => List(Box::new($deser_type(buf)?)),
-                0x0021 => Map(Box::new($deser_type(buf)?), Box::new($deser_type(buf)?)),
-                0x0022 => Set(Box::new($deser_type(buf)?)),
-                0x0030 => {
-                    let keyspace_name =
-                        $read_string(buf).map_err(CqlTypeParseError::UdtKeyspaceNameParseError)?;
-                    let type_name =
-                        $read_string(buf).map_err(CqlTypeParseError::UdtNameParseError)?;
-                    let fields_size: usize = types::read_short(buf)
-                        .map_err(|err| CqlTypeParseError::UdtFieldsCountParseError(err.into()))?
-                        .into();
-
-                    let mut field_types: Vec<(Cow<$l, str>, ColumnType)> =
-                        Vec::with_capacity(fields_size);
-
-                    for _ in 0..fields_size {
-                        let field_name =
-                            $read_string(buf).map_err(CqlTypeParseError::UdtFieldNameParseError)?;
-                        let field_type = $deser_type(buf)?;
-
-                        field_types.push((field_name.into(), field_type));
-                    }
-
-                    UserDefinedType {
-                        type_name: type_name.into(),
-                        keyspace: keyspace_name.into(),
-                        field_types,
-                    }
-                }
-                0x0031 => {
-                    let len: usize = types::read_short(buf)
-                        .map_err(|err| CqlTypeParseError::TupleLengthParseError(err.into()))?
-                        .into();
-                    let mut types = Vec::with_capacity(len);
-                    for _ in 0..len {
-                        types.push($deser_type(buf)?);
-                    }
-                    Tuple(types)
-                }
-                id => {
-                    return Err(CqlTypeParseError::TypeNotImplemented(id));
-                }
-            })
+fn deser_type_generic<'frame, 'result, StrT: Into<Cow<'result, str>>>(
+    buf: &mut &'frame [u8],
+    read_string: fn(&mut &'frame [u8]) -> StdResult<StrT, LowLevelDeserializationError>,
+) -> StdResult<ColumnType<'result>, CqlTypeParseError> {
+    use ColumnType::*;
+    let id =
+        types::read_short(buf).map_err(|err| CqlTypeParseError::TypeIdParseError(err.into()))?;
+    Ok(match id {
+        0x0000 => {
+            // We use types::read_string instead of $read_string here on purpose.
+            // Chances are the underlying string is `...DurationType`, in which case
+            // we don't need to allocate it at all. Only for Custom types
+            // (which we don't support anyway) do we need to allocate.
+            // OTOH, the macro argument function deserializes borrowed OR owned string;
+            // here we want to always deserialize borrowed string.
```

Comment:
Please update the comment, as it still mentions the macro.

Reply:
Fixed